React/Testing best practices.md

```diff
@@ -0,0 +1,610 @@
+```
+
+### User events
+
+For mocking events [testing-library/user-event](https://github.com/testing-library/user-event) is used.
+
+```bash
+npm install --save-dev @testing-library/user-event
+```
+
+### Testing hooks
+
+For testing hooks [react-hooks-testing-library](https://github.com/testing-library/react-hooks-testing-library) is used.
+
+```bash
+npm install --save-dev @testing-library/react-hooks
+```
+
+### Add svg support
+
+```bash
+npm install --save-dev jest-svg-transformer
+```
+
+Add `transform` section to `jest.config.js`
+
+```js
+module.exports = {
+  ...
+  transform: {
+    '^.+\\.tsx?$': 'babel-jest',
+    '^.+\\.svg$': 'jest-svg-transformer',
+  },
+}
+```
+
+## Utils
+
+### Mock providers and store
+
+While testing we need to mock various providers like [`<ChakraProvider>`](https://chakra-ui.com/docs/getting-started#setup-provider/). To mock all provider we can create `./__tests__/test-utils.tsx` file where we will export [custom render](https://testing-library.com/docs/react-testing-library/setup/#custom-render) method with all providers.
+
+```tsx
+const AllProviders = ({ children }) => (
+  <ChakraProvider theme={theme}>{children}</ChakraProvider>
+);
+```
+
+Beside providers we need to mock [datx](https://datx.dev/) store
+
+```tsx
+export const StoreMockContext = React.createContext<AppCollection | null>(null);
+
+const withMockStore = (PageComponent: NextPage) => {
+  const WithMockStore: FC = (props) => {
+    const store: AppCollection = new AppCollection();
+
+    return (
+      <StoreMockContext.Provider value={store}>
+        <PageComponent {...props} />
+      </StoreMockContext.Provider>
+    );
+  };
+
+  return WithMockStore;
+};
+```
+
+Now while testing we can wrap a component in `<StoreMockContext.Consumer>` to get access to the store. Store is needed when we need to mock datx model with relationships.
+
+After mocking all providers and store all needs to be exported.
+
+```tsx
+const customRender = (ui: React.ReactElement, options?: any) =>
+  render(ui, { wrapper: withMockStore(AllProviders), ...options });
+
+// re-export everything
+export * from "@testing-library/react";
+
+// override render method
+export { customRender as render };
+```
+
+Now when testing components we won't import `render` from `@testing-library/react` but instead from `__tests__/test-utils.tsx` like this:
+
+```tsx
+import { render } from "__tests__/test-utils";
+```
+
+Or if we add `path` to `tsconfig`:
+
+```json
+{
+  "compilerOptions": {
+    ...
+    "paths": {
+      ...
+      "@test-utils": ["__tests__/test-utils.tsx"]
+    }
+  },
+}
+```
+
+Then we can import render like this:
+
+```tsx
+import { render } from "@test-utils";
+```
+
+### Folder structure
+
+All tests should be defined in the same directory where the file being tested is. There is one exception and that is `pages` folder because Next.js doesn't allow tests in `pages` folder. Thats why we have `__tests__` folder in the root of our application.
```

Comment:
```suggestion
All tests should be defined in the same directory where the file being tested is. There is one exception, and that is the `pages` folder because Next.js doesn't allow tests in the `pages` folder. That's why we have the `__tests__` folder in the root of our application.
```